### PR TITLE
Add voxel test files

### PR DESCRIPTION
**Pull request by bquan0**
This pull request adds all the files used in the voxel test and essentially acts as an "initial commit".

The voxel test consists of a player who can move around 2 radioactive rods and receive a "radiation" reading based on the location of the player in relation to the radioactive rods. Each rod has a voxel map associated with it, and each voxel in this map has a radiation value which is defined as 1/r, where r is the distance to the rod in the horizontal (x-z) plane. 

In terms of important functionalities implemented by code, `RadioactiveRod.gd` imports the voxel map data from the file `points.csv` and stores it in a 2D array, and `Main.gd` calculates what voxel the player is in relative to each radioactive rod and sums up the radiation values received from all radioactive rods. There are some other scripts which are related to the player and camera movement. 

The files with scripts in them have the `.gd` file extension. The `.tscn` files are the scene files and describe the position of objects in the game.

### CODE REVIEW

File: icon.png
Comment (gonuke) [inline icon.png:1]: Is this file necessary?

File: fonts/Montserrat-Medium.ttf
Comment (gonuke) [inline fonts/Montserrat-Medium.ttf:1]: Is this file necessary?

File: CameraJoint.gd [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+
+var mouse_sensitivity = 0.5
+var scroll_speed = 40
+
+
+# Called when the node enters the scene tree for the first time.
+func _ready():
+	set_as_toplevel(true)
+	spring_length = 7
```
Comment (gonuke) [inline CameraJoint.gd]: (Q based on my language ignorance:) Is there a reason some variables are initialized above at global scope and some here?

File: Main.gd [hunk 2]
```diff
@@ -0,0 +1,54 @@
+extends Node
+
+var rods # all the radioactive rods in the scene
```
Comment (gonuke) [inline Main.gd]: To help move this towards a more general case, why don't we replace `rods` with `sources` and `radiation_array` with `radiation_map`?

File: Main.gd [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+		if !(index < 0 or index > 6858):
+			value = $RadioactiveRod.radiation_array[index][0]
+			voxel_position.x = $RadioactiveRod.radiation_array[index][1]
+			voxel_position.z = $RadioactiveRod.radiation_array[index][2]
+			voxel_position.y = $RadioactiveRod.radiation_array[index][3]
+		# add text for the voxel position the player is in
+		voxel_text += "\nvoxel: (%s," % voxel_position.x
+		voxel_text += "%s," % voxel_position.y
+		voxel_text += "%s)" % voxel_position.z
+		radiation_value += value
```
Comment (gonuke) [inline Main.gd]: Let's make this a lookup function based on `(x,y,z)`.  It can return values and the viz update should happen outside the function.

File: Main.gd [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	for rod in rods:
+		# find player x,y,z with respect to the rod
+		var x = floor(rod.center_x + player_translation.x - rod.translation.x)
+		var y = floor(rod.center_y + player_translation.y - rod.translation.y)
+		var z = floor(rod.center_z + player_translation.z - rod.translation.z)
+	
+		# get the data of the voxel we are in
+		var index = -1
+		# check if we are in the voxel map
+		if x >= 0 and x <= 19 and y >= 0 and y <= 19 and z >= 0 and z <= 19:
```
Comment (gonuke) [inline Main.gd]: Magic numbers - let's make the mesh size a variable that we might learn on load time

File: Main.gd [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+			index = (y * 19 * 19) + (z * 19) + x
+		var value = 0
+		var voxel_position = Vector3.ZERO
+		# only access the array for voxel info if we are in the voxel map
+		if !(index < 0 or index > 6858):
```
Comment (gonuke) [inline Main.gd]: Magic number as above

File: Main.gd [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	# prepend the radiation value 
+	$UserInterface/Label.text = "Radiation: %s" % radiation_value
+	$UserInterface/Label.text += voxel_text
+	
+#	$UserInterface/Label.text += "\nplayer: (%s," % $Player.translation.x
+#	$UserInterface/Label.text += "%s," % $Player.translation.y
+#	$UserInterface/Label.text += "%s)" % $Player.translation.z
+#	$UserInterface/Label.text += "\nrod: (%s," % $RadioactiveRod.translation.x
+#	$UserInterface/Label.text += "%s," % $RadioactiveRod.translation.y
+#	$UserInterface/Label.text += "%s)" % $RadioactiveRod.translation.z
```
Comment (gonuke) [inline Main.gd]: Delete commented lines

File: CameraJoint.gd [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+	spring_length = clamp(spring_length, 0, 12)
+
+func _unhandled_input(event):
+	# move the camera if the RMB is held
+	if event is InputEventMouseMotion and Input.is_action_pressed("right_mouse_button"):
+		rotation_degrees.y -= event.relative.x * mouse_sensitivity
+		rotation_degrees.x = clamp(rotation_degrees.x, -90.0, 90.0)
+		
+		rotation_degrees.x -= event.relative.y * mouse_sensitivity
+		rotation_degrees.y = wrapf(rotation_degrees.y, 0.0, 360.0)
```
Comment (gonuke) [inline CameraJoint.gd]: This looks like it might be a standard pattern/behavior, but it's not intuitive.  Can you add comments and/or make it a function with a clear name?

File: Main.tscn
Comment (gonuke) [inline Main.tscn:1]: As we discussed in person - curious about best practices regarding file directory structure.  My instinct is to put the scene files in a different directory from the script files, but that may be contrary to community practices in the Godot community.

The reason it makes sense to me is that we don't expect there to be review of the scene files - you reported that they are GUI generated.  I suppose they could be hand-generated by an expert and/or machine-generated?

The same goes for other assets like icons, textures, etc.

File: RadioactiveRod.gd
Comment (gonuke) [inline RadioactiveRod.gd:1]: It seems like it makes sense to have the lookup function that I've advised above be inside this file and closely bound to this kind of radiation source.

File: RadioactiveRod.gd [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+	csv_file.open("res://points.csv", File.READ)
+	csv_file.get_csv_line() # skip past first line of labels
+	
+	# copy all values of the csv file into an array
+	while !csv_file.eof_reached(): 
+		var csv_line = csv_file.get_csv_line()
+		var block = []
+		for i in range(0, csv_line.size()):
+			block.append(float(csv_line[i]))
+		radiation_array.append(block)
```
Comment (gonuke) [inline RadioactiveRod.gd]: The way you are currently using/querying this data, don't you just need the first value from each row?

File: RadioactiveRod.gd [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+	
+	csv_file.close()
+	print(radiation_array[0])
```
Comment (gonuke) [inline RadioactiveRod.gd]: This makes sense for debugging, but probably shouldn't commit it

File: RadioactiveRod.gd [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+# Called every frame. 'delta' is the elapsed time since the previous frame.
+#func _process(delta):
+#	pass
```
Comment (gonuke) [inline RadioactiveRod.gd]: Remove commented code

File: icon.png.import
Comment (gonuke) [inline icon.png.import:1]: Does this need to be in repo or is it autogenerated?

File: points.csv.import
Comment (gonuke) [inline points.csv.import:1]: Does this need to be in repo or is it autogenerated?

File: project.godot [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+[input]
+
+move_left={
+"deadzone": 0.5,
+"events": [ Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":0,"alt":false,"shift":false,"control":false,"meta":false,"command":false,"pressed":false,"scancode":65,"physical_scancode":0,"unicode":0,"echo":false,"script":null)
+ ]
+}
+move_right={
+"deadzone": 0.5,
+"events": [ Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":0,"alt":false,"shift":false,"control":false,"meta":false,"command":false,"pressed":false,"scancode":68,"physical_scancode":0,"unicode":0,"echo":false,"script":null)
```
Comment (gonuke) [inline project.godot:29]: This looks like a good place to make a template variable and then copy it and set only the parts that change?  Looks like only the `scancode` changes in each of these.

**Review by gonuke — CHANGES_REQUESTED** [review]
This is a great start and I can see a lot of potential.

File: Main.tscn
Comment (bquan0) [reply Main.tscn:1]: I've looked at a couple of [forums](https://godotforums.org/d/33445-proper-way-of-organizing-project-files/8) and people seem to suggest to keep the scene and related script file in one folder, and have a folder for each scene. However, there doesn't seem to be an official recommendation on how to organize the file structure. Thus, I think I'll try separating the files into a `scenes` and a `scripts` folder as you recommended, and if it gets unmanageable in the future I will change it.

File: icon.png
Comment (bquan0) [reply icon.png:1]: This one isn't necessary and I'll be removing it

File: fonts/Montserrat-Medium.ttf
Comment (bquan0) [reply fonts/Montserrat-Medium.ttf:1]: Technically no, because it is possible to have text without fonts. However, without a font, I'm not able to change the size of the text, which means it will appear very small and be difficult to read.

File: CameraJoint.gd (hunk 1 above)
Comment (bquan0) [reply CameraJoint.gd]: The variables initialized above (`mouse_sensitivity`, `scroll_speed`) are variables that were defined by me to be used in this `CameraJoint.gd` file. `spring_length` is a property of the `CameraJoint` object that I'm setting when the scene is initialized (not defined by me, so I can't put it above)

File: CameraJoint.gd (hunk 7 above)
Comment (bquan0) [reply CameraJoint.gd]: It took me awhile to revisit [where I got this code from](https://youtu.be/UpF7wm0186Q?t=290) to figure out what it's doing, but I finally figured it out. The thing that makes this more confusing is that lines 28 and 31 should be swapped. 

Anyway, the idea here is that changing `rotation_degrees.y` means rotating around the y-axis (vertical axis), which means that this will rotate the camera left or right. Then, we need to make the new y rotation be within the bounds of 0 and 360 degrees (line 31), wrapping it so the player can rotate the camera around the player infinitely without having to go the other way when they reach 0 or 360. It's the same concept for `rotation_degrees.x` except that we cap the rotation to +90 and -90 so the player can't rotate the camera so far up or down (in the up-down rotation we don't want players to infinitely rotate the camera since that could cause an upside down view)

I will add comments relating to this matter.

File: RadioactiveRod.gd (hunk 8 above)
Comment (bquan0) [reply RadioactiveRod.gd]: Yes, but in the future where we have radiation maps that don't follow the assumptions I've made (each voxel is a 1x1x1 cube, the radiation map is sorted in such a way that fits how I find the indices of voxels), I will need the bounds of the voxels to check if the player's position is inside the voxel.

File: points.csv.import
Comment (bquan0) [reply points.csv.import:1]: It's autogenerated, so it can be removed. According to [this forum](https://stackoverflow.com/questions/54249305/godot-what-are-import-files-and-should-you-commit-them-to-git), having them in the repo means less time needed to reimport assets when checking out on a new computer (at the cost of the import files taking up space in the repo)

File: project.godot (hunk 11 above)
Comment (bquan0) [reply project.godot:29]: This isn't something that I access via a coding environment, similar to the `.tscn` files. I think changes are made to this file whenever I modify things in the project.

File: CameraJoint.gd (hunk 7 above)
Comment (gonuke) [reply CameraJoint.gd]: Swapping those lines will probably help the readability too

File: points.csv.import
Comment (gonuke) [reply points.csv.import:1]: Is there a risk that they could become out of sync? Someone updates the human generated file but somehow forgets to update the autogenerated file?  This is one reason to keep autogenerated files out of the repo.

**Comment by bquan0** [comment]
Thanks for the review @gonuke! I just made the suggested changes and answered some questions in a few of the suggestions.

File: scripts/CameraJoint.gd [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+		# note: changing rotation_degrees.x means rotating around the x-axis, 
+		# 		which would be rotating the camera up or down. 
+		#		changing rotation_degrees.y means rotating around the y-axis,
+		#		which would be rotating the camera left or right
+		
+		# rotate the camera left or right
+		rotation_degrees.y -= event.relative.x * mouse_sensitivity
+		rotation_degrees.y = wrapf(rotation_degrees.y, 0.0, 360.0)
+		
+		# rotate the camera left or right
```
Comment (gonuke) [inline scripts/CameraJoint.gd]:
```suggestion
		# rotate the camera up or down
```

File: scripts/RadioactiveRod.gd [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+	# because we define the center to NOT be at (0,0,0), we have to add the
+	# center as an offset.
+	x = floor(center_x + x - translation.x)
+	y = floor(center_y + y - translation.y)
+	z = floor(center_z + z - translation.z)
+
+	var index = -1
+	# check if we are in the voxel map
+	if x >= 0 and x <= max_index_x and y >= 0 and y <= max_index_y and z >= 0 \
+		and z <= max_index_z:
```
Comment (gonuke) [inline scripts/RadioactiveRod.gd]: For readability:
```suggestion
	if x >= 0 and x <= max_index_x and \
	        y >= 0 and y <= max_index_y and \
	        z >= 0 and z <= max_index_z:
```

File: scripts/RadioactiveRod.gd [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+		index = (y * int(pow(max_index_y, 2))) + (z * max_index_z) + x
```
Comment (gonuke) [inline scripts/RadioactiveRod.gd]: cheaper and simpler to just multiply instead of raise to power
```suggestion
		index = (y * max_index_y * max_index_y) + (z * max_index_z) + x
```

File: Main.gd (hunk 4 above)
Comment (bquan0) [reply Main.gd]: Made the mesh size into a variable, but didn't edit `points.csv` to make the program load the mesh size when the game is booted up. I can add that functionality a bit later though, when I know more about the format of the `.csv` files of the radiation maps.

File: scripts/RadioactiveRod.gd [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+	var value = 0
+	var voxel_position = Vector3.ZERO
+	# only access the array for voxel info if we are in the voxel map
+	if !(index < 0 or index >= max_index_x * max_index_y * max_index_z):
```
Comment (gonuke) [inline scripts/RadioactiveRod.gd]: we don't often focus on performance, but this will be called often.  I recommend calculating this once at initialization and storing it as `max_index` or something similar.

**Review by gonuke — CHANGES_REQUESTED** [review]
This definitely does a great job of communicating a path forward to more complex scenarios!  Just a few little final comments.

File: points.csv.import
Comment (bquan0) [reply points.csv.import:1]: Not sure about sync issues. I think it's best to keep the `.import` files out of the repo as we've discussed. I've removed this file from the repo and added `*.import` to the `.gitignore` file.

File: scripts/RadioactiveRod.gd (hunk 15 above)
Comment (bquan0) [reply scripts/RadioactiveRod.gd]: Sounds good. Added these changes in the latest commit

**Comment by gonuke** [comment]
thanks @bquan0 - this looks good

**Review by gonuke — APPROVED** [review]
